src/capture/capture.cpp

```diff
@@ -185,80 +314,166 @@ void CAPTURE_StartVideoCapture()
 void CAPTURE_StopVideoCapture()
 {
 #if (C_SSHOT)
-	if (capturing_video) {
+	switch (capture.state.video) {
+	case CaptureState::Off:
+		LOG_WARNING("CAPTURE: Not capturing video output");
+		break;
+	case CaptureState::Pending:
+		// It's very hard to hit this branch; handling it for
+		// completeness only
+		LOG_MSG("CAPTURE: Cancelling pending video output capture");
+		capture.state.video = CaptureState::Off;
+		break;
+	case CaptureState::InProgress:
 		capture_video_finalise();
-		capturing_video = false;
+		capture.state.video = CaptureState::Off;
 		LOG_MSG("CAPTURE: Stopped capturing video output");
-	} else {
-		LOG_WARNING("CAPTURE: Not capturing video output");
 	}
 #else
 	LOG_WARNING(NoAviSupportMessage);
 #endif
 }
 
-void CAPTURE_AddFrame([[maybe_unused]] const uint16_t width,
-                      [[maybe_unused]] const uint16_t height,
-                      [[maybe_unused]] const uint8_t bits_per_pixel,
-                      [[maybe_unused]] const uint16_t pitch,
-                      [[maybe_unused]] const uint8_t capture_flags,
-                      [[maybe_unused]] const float frames_per_second,
-                      [[maybe_unused]] const uint8_t* image_data,
-                      [[maybe_unused]] const uint8_t* palette_data)
-{
 #if (C_SSHOT)
-	auto image_height = height;
-	if (capture_flags & CaptureFlagDoubleHeight) {
-		image_height *= 2;
+static void handle_add_frame_image_capture(const RenderedImage& image)
+{
+	// No new image capture requests until we finish queuing the current
+	// grouped capture request, otherwise we can get into all sorts of race
+	// conditions.
+	if (capture.state.image.grouped == CaptureState::InProgress) {
+		return;
 	}
-	auto image_width = width;
-	if (capture_flags & CaptureFlagDoubleWidth) {
-		image_width *= 2;
+
+	bool capture_raw      = false;
+	bool capture_upscaled = false;
+	bool capture_rendered = false;
+
+	if (capture.state.image.grouped == CaptureState::Off) {
+		// We're in regular single image capture mode
+		capture_raw = capture.state.image.raw != CaptureState::Off;
+		capture_upscaled = capture.state.image.upscaled != CaptureState::Off;
+		capture_rendered = capture.state.image.rendered != CaptureState::Off;
+
+		// Clear the state flags
+		capture.state.image.raw      = CaptureState::Off;
+		capture.state.image.upscaled = CaptureState::Off;
+		// The `rendered` state is cleared in the
+		// CAPTURE_AddPostRenderImage callback
+	} else {
+		assert(capture.state.image.grouped == CaptureState::Pending);
+		capture.state.image.grouped = CaptureState::InProgress;
+
+		if (grouped_image_capture_formats.capture_raw) {
+			capture_raw = true;
+		}
+		if (grouped_image_capture_formats.capture_upscaled) {
+			capture_upscaled = true;
+		}
+		if (grouped_image_capture_formats.capture_rendered) {
+			capture_rendered = true;
+			// If `rendered` is enabled, the state is cleared in the
```

Comment:
Same with these members:
 - grouped_image_capture_formats.capture_raw
 - grouped_image_capture_formats.capture_upscaled
 - grouped_image_capture_formats.capture_rendered
 
Suggestion (if these are requests or preferences):

 - grouped_image_capture_formats.wants_raw
 - grouped_image_capture_formats.wants_upscaled
 - grouped_image_capture_formats.wants_rendered

Reply:
Yeah I guess I wanted to cut down a bit on the length 😅 They'll be shorter when I convert the whole thing to a class, then I'll add the `is_` prefix.